File: backend/app/schemas/analytics.py
"""Analytics Schemas for TRAP Placement Preparation Operating System.

Defines Pydantic response models for user preparation analytics, problem breakdowns,
company application funnels, goal progress distributions, and study trends.
"""

from typing import Dict, List, Optional, Any
from pydantic import BaseModel, ConfigDict


class ProblemDifficultyStats(BaseModel):
    """Counts of solved problems categorized by difficulty level."""
    easy: int = 0
    medium: int = 0
    hard: int = 0


class ProblemPlatformStats(BaseModel):
    """Problems count grouped by coding platform."""
    platform: str
    count: int


class ProblemTopicStats(BaseModel):
    """Problems count grouped by topic."""
    topic: str
    count: int


class ProblemsAnalytics(BaseModel):
    """Comprehensive problem solving metrics."""
    total_solved: int
    total_tracked: int
    by_difficulty: ProblemDifficultyStats
    by_platform: List[ProblemPlatformStats]
    by_topic: List[ProblemTopicStats]


class CompanyFunnelStats(BaseModel):
    """Company job applications progression stages."""
    wishlist: int = 0
    applied: int = 0
    interviewing: int = 0
    offered: int = 0
    rejected: int = 0


class CompaniesAnalytics(BaseModel):
    """Company tracking and application pipeline analytics."""
    total_tracked: int
    funnel: CompanyFunnelStats
    by_role: Dict[str, int]
    recent_applications: List[Dict[str, Optional[str]]]


class GoalTopicStatusBreakdown(BaseModel):
    """Topic counts categorized by preparation status across all user goals."""
    not_started: int = 0
    bookmarked: int = 0
    in_progress: int = 0
    needs_revision: int = 0
    completed: int = 0
    mastered: int = 0
    skipped: int = 0


class CategoryProgressItem(BaseModel):
    """Progress statistics for a specific category across goals."""
    name: str
    completed_topics: int
    total_topics: int
    percentage: int
    estimated_hours_remaining: float


class GoalsAnalytics(BaseModel):
    """Goal and roadmap completion metrics."""
    total_goals: int
    active_goals: int
    completed_goals: int
    average_progress: int
    total_topics: int
    completed_topics: int
    topic_status_breakdown: GoalTopicStatusBreakdown
    category_progress: List[CategoryProgressItem]


class ActivityTrendItem(BaseModel):
    """Daily problem activity counts for study trend charting."""
    date: str
    problems_solved: int


class StudyDistributionItem(BaseModel):
    """Remaining estimated hours per subject category."""
    category: str
    estimated_hours_remaining: float
    topics_count: int


class ReadinessAnalytics(BaseModel):
    """Placement readiness and category strengths summary."""
    placement_readiness_index: int
    strengths: List[str]
    needs_improvement: List[str]
    current_streak: int
    remaining_study_time: float


class AnalyticsSummaryResponse(BaseModel):
    """Root response model returned by GET /api/v1/analytics/summary."""
    problems: ProblemsAnalytics
    companies: CompaniesAnalytics
    goals: GoalsAnalytics
    readiness: ReadinessAnalytics
    activity_trends: List[ActivityTrendItem]
    study_distribution: List[StudyDistributionItem]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/services/api/analytics.ts
/**
 * Analytics API Service for TRAP Placement Preparation Operating System.
 * 
 * Fetches real analytical aggregations from backend endpoints including problems breakdown,
 * application pipeline funnel, goal progress distributions, and study trends.
 */

import { apiClient } from './client'

export interface ProblemDifficultyStats {
  easy: number
  medium: number
  hard: number
}

export interface ProblemPlatformStats {
  platform: string
  count: number
}

export interface ProblemTopicStats {
  topic: string
  count: number
}

export interface ProblemsAnalytics {
  total_solved: number
  total_tracked: number
  by_difficulty: ProblemDifficultyStats
  by_platform: ProblemPlatformStats[]
  by_topic: ProblemTopicStats[]
}

export interface CompanyFunnelStats {
  wishlist: number
  applied: number
  interviewing: number
  offered: number
  rejected: number
}

export interface CompaniesAnalytics {
  total_tracked: number
  funnel: CompanyFunnelStats
  by_role: Record<string, number>
  recent_applications: Array<{
    name: string
    role: string
    status: string
    applied_date: string | null
  }>
}

export interface GoalTopicStatusBreakdown {
  not_started: number
  bookmarked: number
  in_progress: number
  needs_revision: number
  completed: number
  mastered: number
  skipped: number
}

export interface CategoryProgressItem {
  name: string
  completed_topics: number
  total_topics: number
  percentage: number
  estimated_hours_remaining: number
}

export interface GoalsAnalytics {
  total_goals: number
  active_goals: number
  completed_goals: number
  average_progress: number
  total_topics: number
  completed_topics: number
  topic_status_breakdown: GoalTopicStatusBreakdown
  category_progress: CategoryProgressItem[]
}

export interface ActivityTrendItem {
  date: string
  problems_solved: number
}

export interface StudyDistributionItem {
  category: string
  estimated_hours_remaining: number
  topics_count: number
}

export interface ReadinessAnalytics {
  placement_readiness_index: number
  strengths: string[]
  needs_improvement: string[]
  current_streak: number
  remaining_study_time: number
}

export interface AnalyticsSummary {
  problems: ProblemsAnalytics
  companies: CompaniesAnalytics
  goals: GoalsAnalytics
  readiness: ReadinessAnalytics
  activity_trends: ActivityTrendItem[]
  study_distribution: StudyDistributionItem[]
}

export const analyticsApi = {
  /**
   * Fetch complete analytics summary.
   */
  getSummary: async (): Promise<AnalyticsSummary> => {
    const response = await apiClient.get<AnalyticsSummary>('/analytics/')
    return response.data
  },

  /**
   * Fetch problem solving metrics only.
   */
  getProblems: async (): Promise<ProblemsAnalytics> => {
    const response = await apiClient.get<ProblemsAnalytics>('/analytics/problems')
    return response.data
  },

  /**
   * Fetch company application funnel analytics.
   */
  getCompanies: async (): Promise<CompaniesAnalytics> => {
    const response = await apiClient.get<CompaniesAnalytics>('/analytics/companies')
    return response.data
  },

  /**
   * Fetch goal progress and coverage analytics.
   */
  getGoals: async (): Promise<GoalsAnalytics> => {
    const response = await apiClient.get<GoalsAnalytics>('/analytics/goals')
    return response.data
  },

  /**
   * Fetch placement readiness and strengths.
   */
  getReadiness: async (): Promise<ReadinessAnalytics> => {
    const response = await apiClient.get<ReadinessAnalytics>('/analytics/readiness')
    return response.data
  },
}
